### PR TITLE
vim-patch:8.2.{1727,2324,2363}: charcol(), getcharpos(), getcursorcharpos(), setcharpos(), setcursorcharpos()

### PR DESCRIPTION
**Pull request by zeertzjq**
#### vim-patch:8.2.1727: a popup created with "cursorline" will ignore "firstline"

Problem:    A popup created with "cursorline" will ignore "firstline".
Solution:   When both "cursorline" and "firstline" are present put the cursor
            on "firstline".  Add the "winid" argument to
            getcurpos().
https://github.com/vim/vim/commit/99ca9c4868bb1669706b9e3de9a9218bd11cc459

Skip popup window related code.
Cherry-pick all of Test_getcurpos_setpos() from patch 8.2.0610.

#### vim-patch:8.2.2324: not easy to get mark en cursor posotion by character count

Problem:    Not easy to get mark en cursor posotion by character count.
Solution:   Add functions that use character index. (Yegappan Lakshmanan,
            closes vim/vim#7648)
https://github.com/vim/vim/commit/6f02b00bb0958f70bc15534e115b4c6dadff0e06

#### vim-patch:8.2.2363: curpos() does not accept a string argument as before

Problem:    curpos() does not accept a string argument as before.
solution:   Make a string argument work again. (Yegappan Lakshmanan,
            closes vim/vim#7690
https://github.com/vim/vim/commit/9ebcf231bdccc1673cc92b20f5190fc577ad29d0

#### vim-patch:partial:2346a6378483

Update runtime files
https://github.com/vim/vim/commit/2346a6378483c9871016f9fc821ec5cbea638f13

#### vim-patch:partial:6aa57295cfbe

Update runtime files
https://github.com/vim/vim/commit/6aa57295cfbe8f21c15f0671e45fd53cf990d404

### CODE REVIEW

**Comment by zeertzjq** [comment]
@bfredl Is it fine to consider the remaining code from Vim patch 8.2.1727 as N/A?

**Comment by bfredl** [comment]
@zeertzjq yes, looks good to me.

File: src/nvim/eval.c [hunk 1]
```diff
@@ -8151,6 +8151,58 @@ char *save_tv_as_string(typval_T *tv, ptrdiff_t *const len, bool endnl)
...
+  }
+
+  if (lnum > buf->b_ml.ml_line_count) {
+    lnum = buf->b_ml.ml_line_count;
+  }
+
+  char_u *str = ml_get_buf(buf, lnum, false);
+  if (str == NULL) {
+    return -1;
+  }
```
[the remark this answers is not in the document]
Comment (zeertzjq) [reply src/nvim/eval.c:8170]: I cannot infer this directly, but I found that many places in Vim use the result of `ml_get_buf()` without checking if it is `NULL`.

[the remark this answers is not in the document]
Comment (zeertzjq) [reply src/nvim/eval.c]: Looks like such a `NULL` check never appears anywhere else, so I'll remove it.

**Review by seandewar — APPROVED** [review]
LGTM otherwise; let's hope that NULL check wasn't needed 😄

File: src/nvim/eval.lua [hunk 2]
```diff
@@ -311,8 +314,10 @@ return {
     serverstop={args=1},
     setbufline={args=3, base=3},
     setbufvar={args=3, base=3},
+    setcharpos={args={2, 2}, base=2},
```
[the remark this answers is not in the document]
Comment (zeertzjq) [reply src/nvim/eval.lua]: Thanks!

**Pull request merged**